File: DS/SinglyLinkedList.py
class Node:
	#constructor
	def __init__(self, info, link=None):
		self.info = info
		self.link = link

class LinkedList:

	def __init__(self):
		self.head = None

	def insert_at_begining(self, info):
		#create a new node
		newNode = Node(info)
		if self.head != None:
			newNode.link = self.head
			self.head = newNode
		else:
			self.head = newNode

	def insert_at_end(self, info):
		newNode = Node(info)
		if self.head !=None:
			current = self.head
			while current.link !=None:
				current = current.link 
			current.link = newNode
		else:
			self.head = newNode

	def delete_node(self, ele):
		if self.head == None:
			print("List Empty")
			return
		if self.head.info == ele:
			temp = self.head
			self.head = temp.link 
			temp = None 
			return 
		current = self.head
		while current.link != None:
			if current.link.info == ele:
				temp = current.link 
				current.link = temp.link
				temp = None 
				return 
			current = current.link 
		print("Element is not found in our list")

	def display(self):
		if self.head == None:
			print("List Empty")
			return 
		current = self.head 
		while current !=None:
			print(current.info)
			current = current.link

	def search(self, ele):
		if self.head == None:
			print('list is empty')
			return
		current = self.head
		i=0
		while current != None:
			if current.info == ele:
				print("Element is found in list at position ",i)
				return
			current = current.link
			i=i+1

LL = LinkedList()
LL.insert_at_begining(10)
LL.insert_at_begining(5)
LL.display()
print("*******")
LL.insert_at_end(20)
LL.insert_at_end(30)
LL.display()
LL.delete_node(40)
LL.delete_node(20)
LL.display()
LL.search(30)

File: DS/SinglyLinkedList.go
package main

import (
	"errors"
	"fmt"
)


type Node struct{
	Data int
	Next  *Node
}

func New(i int) *Node{
	n:=Node{Data: i,Next: nil}
	return &n
}



type LinkedList struct {
	head  *Node
}

func (l *LinkedList) InsertAtBeginning(i int)error{
	if l==nil{
		return errors.New(" LinkedList Not initialzed")
	}
	data:=New(i)
	if l.head!=nil{
		temp:=l.head
		l.head=data
		data.Next=temp
	}else{
		l.head=data
	}
	return nil
}

func (l *LinkedList) InsertAtEnd(i int)error{
	if l==nil{
		return errors.New(" LinkedList Not initialzed")
	}
	data:=New(i)
	if l.head!=nil{
		temp:=l.head
		for temp.Next!=nil{
			temp=temp.Next
		}
		temp.Next=data
	}else{
		l.head=data
	}
	return nil
}

func (l *LinkedList) DeleteNode(i int)error{
	if l==nil{
		return errors.New(" LinkedList Not initialzed")
	}
	if l.head==nil{
		return errors.New(" LinkedList Empty")
	}
	if l.head.Data==i{
		fmt.Println("Deleting Data from Head ",l.head.Data)
		l.head=nil
		return nil
	}
	data:=l.head
	for data.Next!=nil{
		if data.Next.Data==i{
			temp:=data.Next
			data.Next=temp.Next
			temp=nil
			return nil
		}
		data=data.Next
	}
	fmt.Println("Element not found in list ")
	return nil
}

func (l *LinkedList) Search(i int)error{
	if l==nil{
		return errors.New(" LinkedList Not initialzed")
	}
	if l.head==nil{
		return errors.New(" LinkedList Empty")
	}
	if l.head.Data==i{
		fmt.Println("Element Found ")
		return nil
	}
	data:=l.head
	for data.Next!=nil{
		if data.Next.Data==i{
			fmt.Println("Element Found ")
			return nil
		}
		data=data.Next
	}
	fmt.Println("Element not found in list ")
	return nil
}


func (l *LinkedList) display()error{
	if l==nil{
		return errors.New(" LinkedList Not initialzed")
	}

	if l.head!=nil{
		data:=l.head
		for data!=nil{
			fmt.Println(data.Data)
			data=data.Next
		}
	}else{
		errors.New(" LinkedList Empty")
	}
	return nil
}


func main(){

	var l LinkedList
	l.InsertAtBeginning(12)
	l.InsertAtBeginning(14)
	l.InsertAtBeginning(16)
	l.InsertAtEnd(18)
	err:=l.display()
	if err!=nil{
		fmt.Println(err)
	}
	fmt.Println("==================")
	fmt.Println("Deleting node ")
	l.DeleteNode(14)
	l.display()
	fmt.Println("After Deletion ")
	fmt.Println("==================")
	l.Search(18)

}
